### ISSUE TITLE
Yolov8 tracker tracking a particular class amongst multiple class detections

### ISSUE THREAD

**Issue by PareshKamble** [ultralytics/ultralytics#6603]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi @glenn-jocher and team
Thank you for this amazing repo.
We know that Yolov8 can use Bot-sort and bytetrack trackers to track multiple objects from the detected classes.
However, we didn't see an option to track only a particular class from the multiple detected classes. Can we have a feature where the detector should detect all the classes but the tracker should track only a particular classe(s)?
In short, we want the detections metadata from all the classes as well as the corresponding tracking metadata. However, a  parameter of **classes to track** may be made available.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@PareshKamble hello!

Thank you very much for the positive feedback and for your interest in contributing to the YOLOv8 project!

Regarding your feature request, the ability to track a specific class or subset of classes from the multiple detected classes is indeed a valuable enhancement. The current tracking implementation does not support this directly, but I understand your use case and why this would be beneficial.

The idea of having a "classes to track" parameter is a good one. This parameter could potentially filter the tracker input to only include objects of certain classes, thus only tracking those while still detecting all classes in the scene.

While YOLOv8 has many modalities and features, as you've rightly pointed out, we're always looking to improve. This kind of feedback can lead to new features that enhance the usability and functionality of the project. 

As for integrating this functionality into YOLOv8, I would suggest reviewing our contribution guidelines detailed in the Docs, and you are more than welcome to submit a pull request with this enhancement. The team and community can then review it to assess its viability and impact on the performance and accuracy.

Thanks again for reaching out and being willing to support the YOLOv8 community. We are looking forward to potentially seeing your contribution in the future. 😊

Keep an eye on the repository for updates, and don't hesitate to share any other ideas or questions you may have!

**Comment by gooelepapa** [comment]
Firstly, I apologize for any errors in my English. Please know that it's not my intention to offend anyone with my words.

Hello! I'm facing a similar issue with my project. To address this, I've implemented a custom predict function. I hope my code might offer some insights or ideas for this situation.

```python
import json
import numpy as np
import torch

from ultralytics import YOLO
import ultralytics.utils.ops as ops
from functools import lru_cache
from BYTETracker.byte_tracker import BYTETracker

class metaBox:
    def __init__(self, data, orig_shape, is_track, names) -> None:
        self.data = data
        self.orig_shape = orig_shape
        self.is_track = is_track
        self.names = names
        self.box = []
        for d in self.data:
            self.box.append(bbox(d, self.orig_shape, self.is_track, self.names)) 

    def __iter__(self):
        return iter(self.box)

    def __len__(self):
        return len(self.box)

    def __str__(self) -> str:
        nl = '\n'
        s = f"{nl}Box data:{nl}{self.data},{nl}"\
            f"Is_track:{nl}{self.is_track}"
        return s
class bbox():
    def __init__(self, data, orig_shape, is_track, names) -> None:
        if data.ndim == 1:
            data = data[None, :]
        self.data = data
        self.orig_shape = orig_shape
        self.is_track = is_track
        self._names = names

    @property
    def names(self):
        return self._names

    @property
    def xyxy(self):
        """Return the boxes in xyxy format."""
        return self.data[:, :4]

    @property
    def conf(self):
        """Return the confidence values of the boxes."""
        return self.data[:, -2]

    @property
    def cls(self):
        """Return the class values of the boxes."""
        return self.data[:, -1]

    @property
    def id(self):
        """Return the track IDs of the boxes (if available)."""
        return self.data[:, -3] if self.is_track else None

    @property
    @lru_cache(maxsize=2)  # maxsize 1 should suffice
    def xywh(self):
        """Return the boxes in xywh format."""
        return ops.xyxy2xywh(self.xyxy)

    @property
    @lru_cache(maxsize=2)
    def xyxyn(self):
        """Return the boxes in xyxy format normalized by original image size."""
        xyxy = self.xyxy.clone() if isinstance(self.xyxy, torch.Tensor) else np.copy(self.xyxy)
        xyxy[..., [0, 2]] /= self.orig_shape[1]
        xyxy[..., [1, 3]] /= self.orig_shape[0]
        return xyxy

    @property
    @lru_cache(maxsize=2)
    def xywhn(self):
        """Return the boxes in xywh format normalized by original image size."""
        xywh = ops.xyxy2xywh(self.xyxy)
        xywh[..., [0, 2]] /= self.orig_shape[1]
        xywh[..., [1, 3]] /= self.orig_shape[0]
        return xywh
class Tracker(BYTETracker):
    def __init__(self, tracker_yaml):
        super().__init__(tracker_yaml)

    def specific_tracking(self, detect_data, track_classes):
        track_boxes = []
        confidences = []
        object_classes = []
        track_results = []
        boxes_data = detect_data
        det_result = np.empty((0,6))
        for bbox in boxes_data:
            x1,y1,x2,y2,conf,cls = bbox
            if int(cls) not in track_classes:
                det_result = np.vstack((det_result, bbox))
            else:
                track_boxes.append([x1, y1, x2, y2])
                confidences.append(conf)
                object_classes.append(cls)
        if len(track_boxes) != 0:
            track_results = self.update(np.array(track_boxes), np.array(confidences), np.array(object_classes))
            if len(track_results) != 0:
                track_results = np.delete(track_results, -1, axis=1)
        return det_result, track_results

class YOLOV8Segment:
    def __init__(self, segmentation_weight='yolov8n-seg.pt'):
        self.segmentation_weight = segmentation_weight
        self.detect_model_segmentation = YOLO(segmentation_weight, task='segment')
        self.model_default_class = [i for i in self.detect_model_segmentation.names.keys()]
        self.tracker = None
        self.update_setting()

    def setup_tracker(self, track_yaml):
        """Set up you ByteTracker yaml.

        Args:
            track_yaml (str): File path of ByteTracker.yaml.
        """
        self.tracker = Tracker(track_yaml)

    def update_setting(self):
        from celeryUtils.lock import lock
        with lock('yoloSetting'):
            with open('CONFIG/YOLOV8_setting.json', 'r') as file:
                settings = json.load(file)
        self.det_conf = settings['conf']
        self.min_box = settings['minBox']
        self.max_box = settings['maxBox']

    # Detect Unit
    def predict(self, frame, iou=0.75, half=True, verbose = False, classes : list=None, track_classes : list=None):
        """Predict given classes and track particular classes.

        Args:
            frame (np.array): cv2 image
            iou (float, optional): NMS threshold. Defaults to 0.75.
            half (bool, optional): Use fp16. Defaults to True.
            verbose (bool, optional): Show predict log. Defaults to False.
            classes (list, optional): A list to predict. Defaults to None.
            track_classes (list, optional): A list to track, the track class must be inside classes. Defaults to None.

        Returns:
            list(metaBox), list(metaBox): return two list, first is detection result, second is track result.
        """
        self.det_iou = iou
        self.det_half = half
        self.det_verbose = verbose
        if classes is not None:
            self.det_classes = classes
        else:
            self.det_classes = self.model_default_class
        preds =  self.detect_model_segmentation.predict(source=frame, task='segment', conf=self.det_conf, iou=self.det_iou, half=self.det_half, verbose = self.det_verbose, classes=self.det_classes)
        if track_classes is None:
            detect_result = []
            track_result = []
            for pred in preds:
                p = pred.boxes.cpu().numpy()
                detect_data = p.data
                detect_result.append(metaBox(detect_data, p.orig_shape, p.is_track, pred.names))
                track_result.append([])
        else:
            assert self.tracker is not None, \
                f'Tracker is not define, please call {self.__class__.__name__}.setup_tracker(track_yaml) before call predict function.'
            assert track_classes == list(set(classes).intersection(set(track_classes))), \
                f'track_classes must be inside classes'
            detect_result = []
            track_result = []
            for pred in preds:
                p = pred.boxes.cpu().numpy()
                detect_data = p.data
                det_data, track_data = self.tracker.specific_tracking(detect_data, track_classes)
                detect_result.append(metaBox(det_data, p.orig_shape, p.is_track, pred.names))
                track_result.append(metaBox(track_data, p.orig_shape, True, pred.names))
        return detect_result, track_result

    def exit(self):
        del self.model
        torch.cuda.empty_cache()



if __name__ == '__main__':
    import cv2
    model = YOLOV8Segment()
    model.setup_tracker('bytetrack.yaml')
    # cap = cv2.VideoCapture("video path")
    # ret, frame = cap.read()
    # if not ret:
    #   model.predict(frame, track_classes=[0])
```
The metaBox object aims to replicate the output format of the Ultralytics Results.
This ensures that everyone can obtain prediction results consistent with the original function.


Additionally, here is the ByteTrack code, which was cloned from [here](<https://github.com/jahongir7174/ByteTrack>): 
```python
import numpy
import yaml
from BYTETracker import util

class State:
    New = 0
    Tracked = 1
    Lost = 2
    Removed = 3


class Track:
    count = 0
    shared_kalman = util.KalmanFilterXYAH()

    def __init__(self, tlwh, score, cls):

        # wait activate
        self._tlwh = numpy.asarray(self.tlbr_to_tlwh(tlwh[:-1]), dtype=numpy.float32)
        self.kalman_filter = None
        self.mean, self.covariance = None, None
        self.is_activated = False

        self.score = score
        self.tracklet_len = 0
        self.cls = cls
        self.idx = tlwh[-1]

    def predict(self):
        mean_state = self.mean.copy()
        if self.state != State.Tracked:
            mean_state[7] = 0
        self.mean, self.covariance = self.kalman_filter.predict(mean_state, self.covariance)

    @staticmethod
    def multi_predict(tracks):
        if len(tracks) <= 0:
            return
        multi_mean = numpy.asarray([st.mean.copy() for st in tracks])
        multi_covariance = numpy.asarray([st.covariance for st in tracks])
        for i, st in enumerate(tracks):
            if st.state != State.Tracked:
                multi_mean[i][7] = 0
        multi_mean, multi_covariance = Track.shared_kalman.multi_predict(multi_mean, multi_covariance)
        for i, (mean, cov) in enumerate(zip(multi_mean, multi_covariance)):
            tracks[i].mean = mean
            tracks[i].covariance = cov

    def activate(self, kalman_filter, frame_id):
        """Start a new tracklet"""
        self.kalman_filter = kalman_filter
        self.track_id = self.next_id()
        self.mean, self.covariance = self.kalman_filter.initiate(self.convert_coords(self._tlwh))

        self.tracklet_len = 0
        self.state = State.Tracked
        if frame_id == 1:
            self.is_activated = True
        self.frame_id = frame_id
        self.start_frame = frame_id

    def re_activate(self, new_track, frame_id, new_id=False):
        self.mean, self.covariance = self.kalman_filter.update(self.mean, self.covariance,
                                                               self.convert_coords(new_track.tlwh))
        self.tracklet_len = 0
        self.state = State.Tracked
        self.is_activated = True
        self.frame_id = frame_id
        if new_id:
            self.track_id = self.next_id()
        self.score = new_track.score
        self.cls = new_track.cls
        self.idx = new_track.idx

    def update(self, new_track, frame_id):
        """
        Update a matched track
        :type new_track: Track
        :type frame_id: int
        :return:
        """
        self.frame_id = frame_id
        self.tracklet_len += 1

        new_tlwh = new_track.tlwh
        self.mean, self.covariance = self.kalman_filter.update(self.mean, self.covariance,
                                                               self.convert_coords(new_tlwh))
        self.state = State.Tracked
        self.is_activated = True

        self.score = new_track.score
        self.cls = new_track.cls
        self.idx = new_track.idx

    def convert_coords(self, tlwh):
        return self.tlwh_to_xyah(tlwh)

    def mark_lost(self):
        self.state = State.Lost

    def mark_removed(self):
        self.state = State.Removed

    @property
    def end_frame(self):
        return self.frame_id

    @staticmethod
    def next_id():
        Track.count += 1
        return Track.count

    @property
    def tlwh(self):
        """Get current position in bounding box format `(top left x, top left y,
        width, height)`.
        """
        if self.mean is None:
            return self._tlwh.copy()
        ret = self.mean[:4].copy()
        ret[2] *= ret[3]
        ret[:2] -= ret[2:] / 2
        return ret

    @property
    def tlbr(self):
        """Convert bounding box to format `(min x, min y, max x, max y)`, i.e.,
        `(top left, bottom right)`.
        """
        ret = self.tlwh.copy()
        ret[2:] += ret[:2]
        return ret

    @staticmethod
    def reset_id():
        Track.count = 0

    @staticmethod
    def tlwh_to_xyah(tlwh):
        """Convert bounding box to format `(center x, center y, aspect ratio,
        height)`, where the aspect ratio is `width / height`.
        """
        ret = numpy.asarray(tlwh).copy()
        ret[:2] += ret[2:] / 2
        ret[2] /= ret[3]
        return ret

    @staticmethod
    def tlbr_to_tlwh(tlbr):
        ret = numpy.asarray(tlbr).copy()
        ret[2:] -= ret[:2]
        return ret

    @staticmethod
    def tlwh_to_tlbr(tlwh):
        ret = numpy.asarray(tlwh).copy()
        ret[2:] += ret[:2]
        return ret

    def __repr__(self):
        return f'OT_{self.track_id}_({self.start_frame}-{self.end_frame})'


class TrackArgs:
    def __init__(self, tracker_yaml) -> None:
        with open(tracker_yaml, 'r') as file:
            self._track_args = yaml.safe_load(file)
    @property
    def track_high_thresh(self):
        return self._track_args['track_high_thresh']
    @property
    def track_low_thresh(self):
        return self._track_args['track_low_thresh']
    @property
    def new_track_thresh(self):
        return self._track_args['new_track_thresh']
    @property
    def track_buffer(self):
        return self._track_args['track_buffer']
    @property
    def match_thresh(self):
        return self._track_args['match_thresh']

class BYTETracker:
    def __init__(self, tracker_yaml):
        self.args = TrackArgs(tracker_yaml)
        self.tracked_tracks = []
        self.lost_tracks = []
        self.removed_tracks = []

        self.frame_id = 0
        # Set args
        self.max_frame_lost = self.args.track_buffer
        self.kalman_filter = util.KalmanFilterXYAH()
        self.reset_id()

    def update(self, boxes, scores, object_classes):
        self.frame_id += 1
        activated_tracks = []
        re_find_tracks = []
        lost_tracks = []
        removed_tracks = []

        # add index
        boxes = numpy.concatenate([boxes, numpy.arange(len(boxes)).reshape(-1, 1)], axis=-1)

        indices_low = scores > self.args.track_low_thresh
        indices_high = scores < self.args.track_high_thresh
        indices_remain = scores > self.args.track_high_thresh

        indices_second = numpy.logical_and(indices_low, indices_high)
        boxes_second = boxes[indices_second]
        boxes = boxes[indices_remain]
        scores_keep = scores[indices_remain]
        scores_second = scores[indices_second]
        cls_keep = object_classes[indices_remain]
        cls_second = object_classes[indices_second]

        detections = self.init_track(boxes, scores_keep, cls_keep)
        """ Add newly detected tracklets to tracked_stracks"""
        unconfirmed = []
        tracked_stracks = []
        for track in self.tracked_tracks:
            if not track.is_activated:
                unconfirmed.append(track)
            else:
                tracked_stracks.append(track)
        """ Step 2: First association, with high score detection boxes"""
        track_pool = self.joint_stracks(tracked_stracks, self.lost_tracks)
        # Predict the current location with KF
        self.multi_predict(track_pool)

        dists = self.get_dists(track_pool, detections)
        matches, u_track, u_detection = util.linear_assignment(dists, thresh=self.args.match_thresh)
        for tracked_i, box_i in matches:
            track = track_pool[tracked_i]
            det = detections[box_i]
            if track.state == State.Tracked:
                track.update(det, self.frame_id)
                activated_tracks.append(track)
            else:
                track.re_activate(det, self.frame_id, new_id=False)
                re_find_tracks.append(track)
        """ Step 3: Second association, with low score detection boxes"""
        # association the untrack to the low score detections
        detections_second = self.init_track(boxes_second, scores_second, cls_second)
        r_tracked_tracks = [track_pool[i] for i in u_track if track_pool[i].state == State.Tracked]
        dists = util.iou_distance(r_tracked_tracks, detections_second)
        matches, u_track, u_detection_second = util.linear_assignment(dists, thresh=0.5)
        for tracked_i, box_i in matches:
            track = r_tracked_tracks[tracked_i]
            det = detections_second[box_i]
            if track.state == State.Tracked:
                track.update(det, self.frame_id)
                activated_tracks.append(track)
            else:
                track.re_activate(det, self.frame_id, new_id=False)
                re_find_tracks.append(track)

        for it in u_track:
            track = r_tracked_tracks[it]
            if track.state != State.Lost:
                track.mark_lost()
                lost_tracks.append(track)
        """Deal with unconfirmed tracks, usually tracks with only one beginning frame"""
        detections = [detections[i] for i in u_detection]
        dists = self.get_dists(unconfirmed, detections)
        matches, u_unconfirmed, u_detection = util.linear_assignment(dists, thresh=0.7)
        for tracked_i, box_i in matches:
            unconfirmed[tracked_i].update(detections[box_i], self.frame_id)
            activated_tracks.append(unconfirmed[tracked_i])
        for it in u_unconfirmed:
            track = unconfirmed[it]
            track.mark_removed()
            removed_tracks.append(track)
        """ Step 4: Init new stracks"""
        for new_i in u_detection:
            track = detections[new_i]
            if track.score < self.args.new_track_thresh:
                continue
            track.activate(self.kalman_filter, self.frame_id)
            activated_tracks.append(track)
        """ Step 5: Update state"""
        for track in self.lost_tracks:
            if self.frame_id - track.end_frame > self.max_frame_lost:
                track.mark_removed()
                removed_tracks.append(track)

        self.tracked_tracks = [t for t in self.tracked_tracks if t.state == State.Tracked]
        self.tracked_tracks = self.joint_stracks(self.tracked_tracks, activated_tracks)
        self.tracked_tracks = self.joint_stracks(self.tracked_tracks, re_find_tracks)
        self.lost_tracks = self.sub_stracks(self.lost_tracks, self.tracked_tracks)
        self.lost_tracks.extend(lost_tracks)
        self.lost_tracks = self.sub_stracks(self.lost_tracks, self.removed_tracks)
        self.removed_tracks.extend(removed_tracks)
        self.tracked_tracks, self.lost_tracks = self.remove_duplicate_stracks(self.tracked_tracks, self.lost_tracks)
        output = [track.tlbr.tolist() + [track.track_id,
                                         track.score,
                                         track.cls,
                                         track.idx] for track in self.tracked_tracks if track.is_activated]
        return numpy.asarray(output, dtype=numpy.float32)

    @staticmethod
    def init_track(boxes, scores, cls):
        return [Track(box, s, c) for (box, s, c) in zip(boxes, scores, cls)] if len(boxes) else []  # detections

    @staticmethod
    def get_dists(tracks, detections):
        dists = util.iou_distance(tracks, detections)
        dists = util.fuse_score(dists, detections)
        return dists

    @staticmethod
    def multi_predict(tracks):
        Track.multi_predict(tracks)

    @staticmethod
    def reset_id():
        Track.reset_id()

    @staticmethod
    def joint_stracks(tlista, tlistb):
        exists = {}
        res = []
        for t in tlista:
            exists[t.track_id] = 1
            res.append(t)
        for t in tlistb:
            tid = t.track_id
            if not exists.get(tid, 0):
                exists[tid] = 1
                res.append(t)
        return res

    @staticmethod
    def sub_stracks(tlista, tlistb):
        stracks = {t.track_id: t for t in tlista}
        for t in tlistb:
            tid = t.track_id
            if stracks.get(tid, 0):
                del stracks[tid]
        return list(stracks.values())

    @staticmethod
    def remove_duplicate_stracks(stracksa, stracksb):
        pdist = util.iou_distance(stracksa, stracksb)
        pairs = numpy.where(pdist < 0.15)
        dupa, dupb = [], []
        for p, q in zip(*pairs):
            timep = stracksa[p].frame_id - stracksa[p].start_frame
            timeq = stracksb[q].frame_id - stracksb[q].start_frame
            if timep > timeq:
                dupb.append(q)
            else:
                dupa.append(p)
        resa = [t for i, t in enumerate(stracksa) if i not in dupa]
        resb = [t for i, t in enumerate(stracksb) if i not in dupb]
        return resa, resb

```

**Comment by glenn-jocher** [comment]
@gooelepapa hello!

Thank you for taking the time to work on a custom predict function and for sharing your code snippet. Due to the complexity and length of the code provided, it's difficult to address specific parts or provide direct feedback here.

If you have identified a way to filter detections for tracking certain classes, that's a great step forward. Please consider summarizing the key changes and how they address the specific issue of tracking selective classes.

Moreover, feel free to contribute to the YOLOv8 project by reviewing our contribution guidelines and submitting a pull request. We highly appreciate contributions from our community that can help enhance YOLOv8.

Keep in mind that maintaining clarity and simplicity in code contributions aids in review and potential integration. I look forward to potentially seeing a distilled version of your custom approach as a contribution to the repository.

Thanks once again! 🚀
